weblate/trans/models/component.py

```diff
@@ -2362,33 +2362,33 @@ def create_translations(
         changed_template: bool = False,
         from_link: bool = False,
         change: int | None = None,
+        run_async: bool = False,
     ) -> bool:
         """Load translations from VCS."""
-        try:
+        if not run_async or settings.CELERY_TASK_ALWAYS_EAGER:
+            # Asynchronous processing not requested or not available, run the update
+            # directly from the request processing.
+            # NOTE: In case the lock cannot be acquired, an error will be raised.
```

Comment:
One question before doing the full review here: Did you change all calls to run async now?

* If yes, then the `run_async` parameter might not be needed anymore.
* If no, then this will now raise `WeblateLockTimeoutError`, but it did fallback to background processing before and this behavior should be restored.

Reply:
Sorry for the late reply, had to dig a bit in the code again ;)

Short answer is: we still need this `run_async` parameter:
* In some cases, I did not feel confident enough about the calling context of `create_translations` to switch it to async (e,g, from some addons, like `MsgmergeAddon`,  `post_commit` function from `git.py`, and other not-directly-related areas like the `new_language` view or `Translation.handle_store_change`).
* We definitively do not want async behavior when calling from the async tasks themselves.

In fact, I wonder if we should not split the `create_translations` in two separate functions:
* `create_translations`: 'public' entry point, essentially the same thing as what is in this PR, with the 'fallback to async' behavior restored. Then we can later check all remaining synchronous calls to it and maybe entirely remove the `run_async` option.
* `create_translations_task`: very thin wrapper around `_create_translations` (just the 'context' lock around it basically), called exclusively from `create_translations`, and the celery tasks code.

Think it would make the general code flow easier to follow and understand, and likely also help in my next goal: to split `_create_translations` itself into smaller bits, and try to create a Celery canvas with them (afaict, once the source language has been updated, it should be possible to update all translations in parallel?).

Reply:
The add-ons should probably stay as it is now (prefer sync, but fallback to async when needed). The reason is that when doing async, it might trigger additional add-on executions which could be avoided when doing that sync.

The `new_language` view is effectively just an optimization to call `create_translations` just once when adding multiple ones, and otherwise it would get called directly via `Component.add_new_language`. That one would need a view change as upon adding user is redirected to the newly created language and expects to see the strings to translate there. These would be missing if `create_translation` did not happen. But in case sync update can not be performed, it is confusing to users even right now.

The `Translation.handle_store_change` is called upon upload which replaces whole file and when deleting a string. The first thing could be changed to async and show the progress, the second should be addressed by https://github.com/WeblateOrg/weblate/issues/10787.

Doing the translations update in parallel is not easily possible because it heavily relies on using shared component object (to cache source string, collect pending source string checks, batch checks). While it would be technically possible to collect that via Celery results, it would complicate the code, and it would probably make the performance worse (for example because of fetching source strings from the database for each translation separately).

Reply:
Thanks for all the info! I'll try to update the PR accordingly this weekend.